### ISSUE TITLE
regex pattern-matching : match implements a search-like behavior, instead of a match

### ISSUE THREAD

**Issue by smunix** [Morgan-Stanley/hobbes#138]
I noted the following issues with regex pattern matching :

1.- binding the same name from 2 different pattern rows confuses the result :
> match "fooooooooobar" with | 'f(?<o>o*)bar' -> o | '(?<h>.+)(?<o>o*)bar' -> o | _ -> "none"
""
> match "fooooooooobar" with | 'f(?<o>o*)bar' -> o | '(?<h>.+)(?<od>o*)bar' -> od | _ -> "none"
"ooooooooo"

2.- `.+` followed by another regex in sequence obliterates the sequenced regexes. FAalthough this behavior is somewhat coorect, a user might expect a stricter matching behaviour to be implemented here. Some regex engines differentiates between matches and searches; is it something we'd benefit from having here too ? 
> match "fooooooooobar" with | 'f(?<o>o*)bar' -> o | '(?<h>.+)(?<od>o*)bar' -> h | _ -> "none"
"ooooooooo"
> match "fooooooooobar" with | 'a(?<o>o*)bar' -> o | '(?<h>.+)(?<od>o*)bar' -> h | _ -> "none"
"fooooooooobar"
> match "fooooooooobar" with | 'a(?<o>o*)bar' -> o | '(?<hd>.+)(?<od>o*)bar' -> hd | _ -> "none"
"fooooooooobar"
> match "fooooooooobar" with | 'a(?<o>o*)bar' -> o | '(?<hd>.)(?<od>o*)bar' -> hd | _ -> "none"
"f"
> match "fooooooooobar" with | 'a(?<o>o*)bar' -> o | '(?<hd>.+)(?<od>o*)bar' -> hd | _ -> "none"
"fooooooooobar"

**Comment by kthielen** [comment]
Yes, good points about the regex binding code.  That support was initially added with just the 2-row either/or match in mind.  It's worth a second look.
